### PR TITLE
Reject webpack, embrace vite

### PR DESCRIPTION
**Pull request by aekasitt**
Aside from the later add-ons now available with Vite + Vue combination, I just want to bring the attention to build-time comparison between the two build-tools: Webpack vs. Vite.

`Webpack Build-time takes around 5 - 6 seconds`

webpack-buildtime

`Vite Build-time took 0.3 seconds`

vite-buildtime

**Notes**
* Maybe missing some electron features but should be an easy fix.
* Added VitePWA plugin to make sure `.webmanifest` file gets created at build-time.
* Had to hard-code productName, productDescription and context override for cordova app.

### CODE REVIEW

**Comment by callebtc** [comment]
Very interesting, opinionated and thorough! I like the initiative!

Does this preserve all build functionalities of Quasar (PWA, SPA, mobile, etc..)?

@ebrakke @cjbeery24 you might be more experienced with these things, what do you think?

**Comment by ebrakke** [comment]
I love the idea of using vite! Though I'm curious why we would get rid of quasar.
There's a guide from quasar about how to use vite instead of webpack: https://quasar.dev/quasar-cli-vite/convert-to-quasar-cli-with-vite

**Comment by ebrakke** [comment]
I would suggest we use this PR: https://github.com/cashubtc/cashu.me/pull/34

**Comment by aekasitt** [comment]
I'll check it out. Didn't remove `quasar` as mentioned though, but `quasar` being used as script command seems a bit out-of-scope for a frontend framework. I always avoid using `vue-cli`, `nuxt-cli`, `quasar` for build-tool to be clear as day when debugging.

+1 on using `vitest` on #34 
That was in the roadmap too if this PR gets a positive feedback.

**Review by aekasitt — COMMENTED** [review]
Resolved merge conflicts.

**Comment by aekasitt** [comment]
**Notes on #34**
- I think `@noble/hashes` and `@noble/secp256k1` usage are needed but should be done with a separate PR with parity tests showing behavior of lib doesn't change. (Or specific changes made if previous lib faulty).
- `vitest` -> `jest` should also be a different PR too imho.

** Notes on this PR (#33)**
- Using comments by @callebtc on #34, I separated build-scripts into two, `build` and `build:pwa` decide whether or not certain assets get included in `dist` build directory. And then can view the differences via `preview` command.
- Improve syntaxes of boot auto-imports.
- Using `vite` build-tool directly via `vite.config.js` as build configs has the added benefits of `resolve.alias` feature as well. This is why I changed a lot of import paths in this PR.
- Using `vite` build-tool directly via `vite.config.js` as build configs has the added benefits of streamlining future `JavaScript -> TypeScript` refactor if needed, including specifying default `import.meta.env.USER_VARS` types beforehand.
- Using `vite` build-tool directly via `vite.config.js` as build configs has the added benefits of streamlining future `Vitest integration` if needed. (Seen on PR #34) seeing plugin imports can be consolidated.

**Comment by callebtc** [comment]
Is history mode on "hash"?
image

**Comment by callebtc** [comment]
Loading the page with a fresh browser opens it in light mode. Dark mode can't be toggled on. Multiple refreshes later it suddenly works! 
image

**Comment by aekasitt** [comment]
Cannot reproduce the Light-mode default nor the Dark-mode toggle-able hurdle on fresh browser.
Attached are both Chrome (left) and Firefox (right) using `localhost:8080` with local data cleared.

chrome-firefox-dark

**Comment by aekasitt** [comment]
`#` in the path is an artifact of vite's hot module reload. (hmr)
Attached is the screen-capture when running `npm run preview` after build. No `#` to be found.
cashu-preview

**Comment by ebrakke** [comment]
My main hesitation with this PR is that I'm unsure that this codebase needs to have a custom vite config as compared with the just using `quasar` as a build tool.

For one, the quasar build tool was not preventing us from doing anything that we needed to do for development. Even with webpack, it's still quite fast to build.

I see the merits of having more control over the build system by using vite directly, but I would say there's no reason to do it until the application has a need that we cannot accomplish otherwise. Because once we move away from quasar, we now have to maintain anything that the quasar CLI was giving us for free.

IMO, there's higher priority things that the app would benefit from (code refactoring, leveraging cashu-ts, decoupling view logic from core business logic, potentially typescript, etc...)

**Comment by aekasitt** [comment]
**Clarification needed on the hesitation.**
#33 and #34 both aim to get rid of webpack, with #34 taking a step further and replacing `jest` with `vitest` as well. Are we saying that `webpack` should stay?

**Why I think we should leave `quasar.config.json` in the dust**
Only two of those options; `plugins` and `extras`, were front-end specific, easily redefined on `src/main.js`. The rest were build configs where quasar does not excel at and only passes on to build-tools underneath, previously `webpack` (PWA manifest included). This abstraction does not add, only obscure.

**Vite's, as a build-tool, immediate benefits**
The added benefits on top of `vite` over using `quasar` build-tools is that `vite` gets to make project structure, (why I'm here), customizable by defining `resolve.alias` as such...

```
import { fileURLToPath, URL } from "node:url";
...
export default defineConfig({
  ...,
  resolve: {
    alias: {
      '~~/': fileURLToPath(new URL("./src/", import.meta.url)),
    }
  }
})
```
**Notes**
- If this was possible with quasar-config, would like to see it done to make project structure more approachable.

---
Other added benefits are for example:
- Seamless integration with `vitest` as seen on #34 
- My favorite plugin: [vite-svg-loader](https://github.com/jpkleemans/vite-svg-loader)
- PWA Plugin: [vite-plugin-pwa](https://github.com/vite-pwa/vite-plugin-pwa)
- Electron plugin: [vite-plugin-electron](https://github.com/caoxiemeihao/vite-electron-plugin)
- Electron Renderer plugin: [vite-plugin-electron-renderer](https://github.com/electron-vite/vite-plugin-electron-renderer)

**Comment by ebrakke** [comment]
I’m just saying if we do vite, I think it makes the most sense to do it with quasar, since this is a quasar app. 
https://quasar.dev/quasar-cli-vite/handling-vite/
You can override the vite config from the quasar config file, so I believe all aliasing can be done there. 
Quasar cli already gives us pwa support, so I don’t see the benefit of changing it to the vite pwa plug-in. 

We just added jest last week, with minimal tests. I only added jest because we had webpack still. I don’t see the change to vitest in #34 as a larger change, as none of the test actually change. 

TL;DR this application is a quasar app, and it seems like the recommended way to build a quasar app is to also use their CLI. I think if / when we need to do something that the CLI doesn’t support, that would be the right time to look into a more custom build approach.

**Comment by aekasitt** [comment]
Saw #34 merged, will close this now.
